New Post - State of CSS 2022 (#7807)
GoogleChrome/web.dev: src/site/content/en/blog/2022-state-of-css/index.md

---
layout: post
title: State of CSS 2022
subhead: Web styling features of today and tomorrow, as seen at Google IO 2022, plus some extras.
authors:
  - adamargyle
description: Web styling features of today and tomorrow, as seen at Google IO 2022, plus some extras.
date: 2022-05-11
thumbnail: image/vS06HQ1YTsbMKSFTIPl2iogUQP73/iOfuipzyniP6aec4cOss.png
hero: image/vS06HQ1YTsbMKSFTIPl2iogUQP73/uT454OnumDCJbV0PWmvu.png
alt: Colorful lettering using a Colr Font saying State of CSS 2022
tags:
  - blog
  - css
  - html
---

The year 2022 is set to be one of CSS's greatest years, in both features and
cooperative browser feature releases, with a collaborative goal to implement 14
features!

## Overview

> This post is the article form of the talk given at Google IO 2022. It's not

Author: nice, yes, great idea. i should have a link to that soon.

time. Then, as styles load, are loaded or defined, they can be placed within a
layer, allowing a preservation of style override importance but without the
meticulously managed loading orchestration.

{% Video
  src="video/vS06HQ1YTsbMKSFTIPl2iogUQP73/MQJ5WhdqY78qTwIU2Bt6.mp4",
  autoplay="true",
  loop="true",
  muted="true"
%}

The video shows how the defined cascade layers allow for a more liberated and
freestyle authoring and loading process, while still maintaining the cascade as
needed.

Chrome DevTools is helpful for visualizing which styles are coming from which
layers:

{% Img 
  src="image/vS06HQ1YTsbMKSFTIPl2iogUQP73/17n44fLAbpo0wd6O6Gae.png", 
  alt="Screenshot of the Styles sidebar of Chrome Devtools, highlighting how styles appear within new Layer groups.", 
  width="771", 
  height="689" 
%}

**Resources**  
Spec: [https://www.w3.org/TR/css-cascade-5/#layering](https://www.w3.org/TR/css-cascade-5/#layering)  
Explainer: [https://css.oddbird.net/layers/explainer/](https://css.oddbird.net/layers/explainer/)   
MDN: [https://developer.mozilla.org/docs/Web/CSS/@layer](https://developer.mozilla.org/docs/Web/CSS/@layer)   

> Post by [Una Kravets](https://twitter.com/Una):  [https://developer.chrome.com/blog/cascade-layers/](https://developer.chrome.com/blog/cascade-layers/)  

Author: i went through a few iterations on this and preferred the presentation of the bolded line and the lack of it's entry into the table of contents. but i agree that it seem header worthy, and if that's the case it should follow the numbering established. i just.. it doesnt look as good 🙂 

Current:
![Screen Shot 2022-04-29 at 8 36 35 AM](https://user-images.githubusercontent.com/1134620/165977808-b5e7d698-8a9a-4c8e-b4e4-60098176cefd.png)

Proposed:
![Screen Shot 2022-04-29 at 8 39 07 AM](https://user-images.githubusercontent.com/1134620/165977828-602c4e32-9045-4b9f-bb62-59f99944f85c.png)

the condensed presentation just for more like citations and resources to me. but if you insist, i'll move all them to a header list!

> Post by [Una Kravets](https://twitter.com/Una):  [https://developer.chrome.com/blog/cascade-layers/](https://developer.chrome.com/blog/cascade-layers/)  

Author: never mind, i'll update all of them!

}
```

Lastly, children of `<nav>` or `<main>` can align or size themselves using the
`fullbleed` and `main` columns and lines.

```css
.main-content {
  grid-column: main;
}

.fullbleed {
  grid-column: fullbleed;
}
```

{% Codepen {
 user: 'web-dot-dev',
 id: 'JjMQzVV',
 tab: 'result'
} %}

Devtools can help you see the lines and subgrids (Firefox only at the moment).
In the following image, the parent grid and subgrids have been overlaid. It now
resembles how designers were thinking about the layout.

{% Img 
  src="image/vS06HQ1YTsbMKSFTIPl2iogUQP73/9V3fShGvncWBFIkfetY3.png", 
  alt="Screenshot of a subgrid demo, using the Chrome Devtools grid overlay tooling to show the lines defined by CSS.", 
  width="800", 
  height="765" 
%}

In the elements panel of devtools you can see which elements are grids and
subgrids, which is very helpful for debugging or validating layout.

<figure data-size="full">
  {% Img 
    src="image/vS06HQ1YTsbMKSFTIPl2iogUQP73/2kDFaCapNPGWL2a8XCwI.png", 
    alt="Screenshot of the Chrome Devtools Elements panel labelling which elements have grid or subgrid layouts.", 
    width="800", 
    height="509" 
  %}
  <figcaption>

>     Screenshot from Firefox Devtools

Reviewer: ```suggestion
    Screenshot from Firefox Devtools.
```

assist designers and developers in displaying, picking, and mixing colors.
Previously, sRGB was the only option for working with color, but now CSS unlocks
new potential and a new default color space, LCH.

#### `color-mix()`

{% BrowserCompat 'css.types.color.color-mix' %}

Before `color-mix()`, developers and designers needed preprocessors like
[Sass](https://sass-lang.com/) to mix the colors before the browser saw them.
Most color-mixing functions also didn't provide the option to specify which
color space to do the mixing in, sometimes resulting in confusing results.

After `color-mix()`, developers and designers can mix colors in the browser,
alongside all their other styles, without running build processes or including
JavaScript. Additionally, they can specify which color space to do the mixing
in, or use the default mixing color space of LCH.

Often, a brand color is used as a base and variants are created from it, such as
lighter or darker colors for hover styles. Here's what that looks like with
`color-mix()`:

```css
.color-mix-example {
  --brand: #0af;

  --darker: color-mix(var(--brand) 25%, black);
  --lighter: color-mix(var(--brand) 25%, white);
}
```

> and if you wanted to mix those colors in a different color space, like srgb,

Reviewer: ```suggestion
To mix those colors in a different color space, like srgb,
```

pick the highest contrasting color from the selection:

```css
color: color-contrast(gray vs indigo, rebeccapurple, hotpink);
```

Lastly, in case it's preferable not to pick the highest contrasting color from
the list, a target contrast ratio can be provided, and the first color to pass
it is chosen:

```css
color: color-contrast(
  var(--bg-blue-1)
  vs
  var(--text-lightest), var(--text-light), var(--text-subdued)
  to AA /* 4.5 could also be passed */
);
```

This function can be used for more than just text color, though I estimate that
will be its primary use case. Think about how much easier it will be to deliver
accessible and legible interfaces once the choosing of proper contrasting colors
is built into the CSS language itself.

**Resources**  
Spec: [https://www.w3.org/TR/css-color-5/#colorcontrast](https://www.w3.org/TR/css-color-5/#colorcontrast)  
MDN: [https://developer.mozilla.org/docs/Web/CSS/color_value/color-contrast()](https://developer.mozilla.org/docs/Web/CSS/color_value/color-contrast())  
Demo: [https://codepen.io/web-dot-dev/pen/qBpzwZW](https://codepen.io/web-dot-dev/pen/qBpzwZW)  

#### Relative color syntax

> <!-- {% BrowserCompat 'css.types.color' %} -->

Reviewer: this might be css.types.color.color

of the page was frozen, or `inert`. Users’ focus is placed inside the alert
dialog and has nowhere else to go. Once the user interacts and completes the
alert function request, the page is interactive again. `inert` empowers
developers to achieve this same guided focus experience with ease.

Here's a small code sample to show how it works:

```css/7
<body>
  <div class="modal">
    <h2>Modal Title</h2>
    <p>...<p>
    <button>Save</button>
    <button>Discard</button>
  </div>
  <main inert>
    <!-- cannot be keyboard focused or clicked -->
  </main>
</body>
```

A dialog is a great example, but `inert` is also helpful for things such as the
slide-out side menu user experience. When a user slides out the side menu, it's
not OK to let the mouse or keyboard interact with the page behind it; that's a
bit tricky for users. Instead, when the side menu is showing, make the page
inert, and now users must close or navigate within that side menu, and won't
ever find themselves lost somewhere else in the page with an open menu.

**Resources**  
Spec: [https://html.spec.whatwg.org/multipage/interaction.html#inert](https://html.spec.whatwg.org/multipage/interaction.html#inert)  

> MDN: [https://developer.mozilla.org/docs/Web/API/HTMLElement/inert](https://developer.mozilla.org/docs/Web/API/HTMLElement/inert)

Reviewer: We can add Emma's post now https://developer.chrome.com/blog/inert/

changed, and any `vh` units would shift and resize as their target size changed.
In later years, the `vh` unit specifically needed to decide which of the two
viewport sizes it was going to use, because it was causing jarring visual layout
issues on mobile devices. It was determined that the `vh` would always represent
the largest viewport.

```css
.original-viewport-units {
  height: 100vh;
  width: 100vw;
  --size: 100vmin;
  --size: 100vmax;
}
```

After the new viewport variants, small, large, and dynamic viewport units are
made available, with the addition of [logical
equivalents](/learn/css/logical-properties/) to the physical ones. The idea is
to give developers and designers the ability to choose which unit they want to
use for their given scenario. Maybe it's ok to have a small jarring layout shift
when the status bar goes away, so then `dvh` (dynamic viewport height) could be
used without worry.

{% Img 
  src="image/vS06HQ1YTsbMKSFTIPl2iogUQP73/KtZ9bXwtzmo7hguwQOWO.png", 
  alt="ALT_TEXT_HERE", 
  width="800", 
  height="664" 
%}

Here's a complete list of all the new viewport unit options made available with
the new viewport variants:

> <div class="table-wrapper scrollbar">

Reviewer: not sure that a table is best for this, if you use one however you can't use markdown inside it.

Author: fixed 👍🏻

during build time.

After `@custom-media`, CSS allows aliasing media queries and the referencing of
them, just like a custom property.

Naming things is very important: it can align purpose with the syntax, making
things easier to share and easier to use in teams. Here are a few custom media
queries that follow me between projects:

```css
@custom-media --OSdark  (prefers-color-scheme: dark);
@custom-media --OSlight (prefers-color-scheme: light);

@custom-media --pointer (hover) and (pointer: coarse);
@custom-media --mouse   (hover) and (pointer: fine);

@custom-media --xxs-and-above (width >= 240px);
@custom-media --xxs-and-below (width <= 240px);
```

Now that they're defined, I can use one of them like this:

```css
@media (--OSdark) {
  :root {
    …
  }
}
```

Find a [full list of custom media
queries](https://open-props.style/#media-queries) I use inside my CSS custom
property library [Open Props](https://open-props.style).

**Resources**  
Spec: [https://www.w3.org/TR/mediaqueries-5/#custom-mq](https://www.w3.org/TR/mediaqueries-5/#custom-mq)  
PostCSS plugin: [https://github.com/postcss/postcss-custom-media](https://github.com/postcss/postcss-custom-media)  

> ### Nesting selectors is so nice

Reviewer: Can we mention PostCSS here? You have it in the links but it would be good as an aside.

Author: yep! done 👍🏻

Spec: [https://www.w3.org/TR/css-scoping-1/](https://www.w3.org/TR/css-scoping-1/)  
Explainer: [https://css.oddbird.net/scope/explainer/](https://css.oddbird.net/scope/explainer/)  

### No CSS way for a masonry layout

Before CSS masonry with grid, JavaScript was the best way to achieve a masonry
layout, as any of the CSS methods with columns or flexbox would inaccurately
represent the content order.

After CSS masonry with grid, no JavaScript libraries will be required and the
content order will be correct.

<figure data-size="full">
  {% Img 
    src="image/vS06HQ1YTsbMKSFTIPl2iogUQP73/xfOiyctwsXa8wnfkuoS2.png", 
    alt="Screenshot of the masonry layout which shows numbers traveling along the top, then going down.", 
    width="800", 
    height="547" 
  %}
  <figcaption>
    Image and demo from Smashing Magazine<br>
    <a href="https://www.smashingmagazine.com/native-css-masonry-layout-css-grid/">https://www.smashingmagazine.com/native-css-masonry-layout-css-grid/</a>
  </figcaption>
</figure>

The preceding demo is achieved with the following CSS:

```css/3
.container {
  display: grid;
  grid-template-columns: repeat(4, 1fr);
  grid-template-rows: masonry;
}
```

> It's still under debate whether or not CSS grid is the best place for this

Reviewer: Not sure this is under debate at this point, I think that ship may have sailed.

Author: cool, will update

  alt="Screenshot of a TV show carousel interface with no thumbnails and many titles shown.", 
  width="800", 
  height="528" 
%}

There are more advantages to this reduced data experience than just data
savings. More titles can be seen and there's no distracting cover images to
steal attention. Many users browse in a data saver mode because they pay per
megabyte of data—it's really nice to see CSS able to help out here.

**Resources**  
Spec: [https://www.w3.org/TR/mediaqueries-5/#prefers-reduced-data](https://www.w3.org/TR/mediaqueries-5/#prefers-reduced-data)  
MDN: [https://developer.mozilla.org/docs/Web/CSS/@media/prefers-reduced-data](https://developer.mozilla.org/docs/Web/CSS/@media/prefers-reduced-data)  
GUI Challenges: [https://gui-challenges.web.app/media-scroller/dist/](https://gui-challenges.web.app/media-scroller/dist/)  
Smashing Magazine: [https://www.smashingmagazine.com/2021/12/core-web-vitals-case-study-smashing-magazine/#savedata-and-prefers-reduced-data](https://www.smashingmagazine.com/2021/12/core-web-vitals-case-study-smashing-magazine/#savedata-and-prefers-reduced-data)  

### Scroll snap features are too limited

> Before these scroll snap proposals, writing your own JavaScript to manage a

Reviewer: ```suggestion
Before the following scroll snap proposals, writing your own JavaScript to manage a
```
